src/pstore/pstore.c

```diff
@@ -0,0 +1,582 @@
+                    LIST_REMOVE(dlcl, (*list), dc);
+                    LIST_APPEND(dlcl, (*list), dc);
+                    sorted = 0;
+                    break; /* re-start sort */
+                }
+            }
+            else
+            {
+                if (!(dc == *list) && /* not start of list */
+                    /* lexigraphical sort, descending */
+                    (strcmp(dc->de.d_name, dc->dlcl_prev->de.d_name) > 0))
+                {
+                    LIST_REMOVE(dlcl, (*list), dc);
+                    LIST_PREPEND(dlcl, (*list), dc);
+                    sorted = 0;
+                    break; /* re-start sort */
+                }
+            }
+        }
+    } while (!sorted);
+}
+
+static void free_files (dirent_ct **list) {
+    dirent_ct *dc;
+    LIST_FOREACH(dlcl, dc, *list) {
+        free(dc->contents);
+        LIST_REMOVE(dlcl, (*list), dc);
+        free(dc);
+        break; /* re-start loop */
+    }
+}
+
+static int move_file (dirent_ct *dc, const char *subdir) {
+    _cleanup_free_ char *ifdpath;
+    int remove_file = 0;
+    int r = 0;
+
+    ifdpath = path_join(arg_sourcedir, dc->de.d_name);
+
+    if (arg_storage == PSTORE_STORAGE_EXTERNAL)
+    {
+        /* This code copies the file from the pstore to other storage.
+       The rename() syscall is not utilized as it results in the
+       Invalid cross-device link error.
+       In addition, an optional subdirectory can be specified in
+       forming the final destination path.
+        */
+        _cleanup_free_ char *ofdpath;
+        ofdpath = path_join(arg_archivedir, subdir, dc->de.d_name);
+
+        /* Make sure destination exists */
+        r = mkdir_parents(ofdpath, S_IRWXU|S_IRGRP|S_IXGRP|S_IROTH|S_IXOTH /*0755*/);
+        if (r == 0)
+        {
+#if 1
+            r = copy_file_atomic(ifdpath, ofdpath, 0644, 0, 0, COPY_REPLACE);
+            remove_file = (r == 0);
+#else
+            int ofd;
+
+            ofd = open(ofdpath, O_CREAT|O_TRUNC|O_WRONLY, S_IRUSR|S_IWUSR|S_IRGRP|S_IROTH/*0644*/);
+            if (ofd < 0)
+            {
+                log_error_errno(errno, "open(%s): %m\n", ofdpath);
+                r = errno;
+            }
+            else
+            {
+                if (dc->contents != NULL)
+                {
+                    ssize_t wr;
+                    wr = write(ofd, dc->contents, dc->statbuf.st_size);
+                    /* Detect write problem */
+                    if (wr != dc->statbuf.st_size)
+                    {
+                        log_error_errno(errno, "write(%s): %m\n", ofdpath);
+                        r = errno;
+                    }
+                    else
+                        remove_file = 1;
+                }
+            }
+            close(ofd);
+// FIX??? update ofd stat.time to same info as dc.stat.time?
+#endif
+        }
+        else
+        {
+            log_error_errno(errno, "mkdir_parents(): %m\n");
+            r = errno;
```

Comment:
log_error_errno() returns the negative error number passed in as first arg back, please use that

Reply:
done